src/main/java/org/zalando/nakadi/repository/MultiTimelineEventConsumer.java

```diff
@@ -0,0 +1,264 @@
+import org.zalando.nakadi.exceptions.NakadiException;
+import org.zalando.nakadi.exceptions.NakadiRuntimeException;
+import org.zalando.nakadi.exceptions.ServiceUnavailableException;
+import org.zalando.nakadi.service.timeline.TimelineService;
+import org.zalando.nakadi.service.timeline.TimelineSync;
+import org.zalando.nakadi.util.NakadiCollectionUtils;
+
+public class MultiTimelineEventConsumer implements EventConsumer {
+    private final String clientId;
+    private final Map<EventTypePartition, NakadiCursor> latestOffsets = new HashMap<>();
+    private final Map<String, List<Timeline>> consumedEventTypes = new HashMap<>();
+    private final Map<String, TimelineSync.ListenerRegistration> timelineRefreshListeners = new HashMap<>();
+    private final Map<TopicRepository, EventConsumer> eventConsumers = new HashMap<>();
+    private final LinkedList<ConsumedEvent> eventsQueue = new LinkedList<>();
+    private final Map<EventTypePartition, String> borderOffsets = new HashMap<>();
+    private final TimelineService timelineService;
+    private final TimelineSync timelineSync;
+    private final BlockingQueue<Runnable> queuedTasks = new LinkedBlockingQueue<>();
+    private static final Logger LOG = LoggerFactory.getLogger(MultiTimelineEventConsumer.class);
+
+    public MultiTimelineEventConsumer(
+            final String clientId,
+            final TimelineService timelineService,
+            final TimelineSync timelineSync) {
+        this.clientId = clientId;
+        this.timelineService = timelineService;
+        this.timelineSync = timelineSync;
+    }
+
+    @Override
+    public Set<TopicPartition> getAssignment() {
+        return latestOffsets.values().stream().map(NakadiCursor::getTopicPartition).collect(Collectors.toSet());
+    }
+
+
+    @Override
+    public Optional<ConsumedEvent> readEvent() {
+        Runnable task;
+        while ((task = queuedTasks.poll()) != null) {
+            task.run();
+        }
+        final Optional<ConsumedEvent> result = poll();
+        if (result.isPresent()) {
+            final NakadiCursor position = result.get().getPosition();
+            final EventTypePartition etp = position.getEventTypePartition();
+            latestOffsets.put(etp, position);
+            final String borderOffset = borderOffsets.get(etp);
+            if (null != borderOffset && borderOffset.compareTo(position.getOffset()) <= 0) {
+                try {
+                    reelectTopicRepositories();
+                } catch (final NakadiException | InvalidCursorException e) {
+                    throw new NakadiRuntimeException(e);
+                }
+            }
+        }
+        return result;
+    }
+
+    private Optional<ConsumedEvent> poll() {
+        Optional<ConsumedEvent> result = Optional.ofNullable(eventsQueue.poll());
+        if (!result.isPresent()) {
+            eventConsumers.values().forEach(consumer -> consumer.readEvent().ifPresent(eventsQueue::add));
```

Comment:
what about reading in parallel? I don't know if that would optimise things since kafka consumer already pre-fetches data, so that when calling `poll` the data is already available. What do you think?